src/wasm-interpreter.h

```diff
@@ -2314,19 +2320,28 @@ template<typename GlobalManager, typename SubType> class ModuleInstanceBase {
 
   void initializeTableContents() {
     ModuleUtils::iterActiveElementSegments(wasm, [&](ElementSegment* segment) {
-      Address offset =
-        (uint32_t)InitializerExpressionRunner<GlobalManager>(globals, maxDepth)
-          .visit(segment->offset)
-          .getSingleValue()
-          .geti32();
-
       Function dummyFunc;
       FunctionScope dummyScope(&dummyFunc, {});
       RuntimeExpressionRunner runner(*this, dummyScope, maxDepth);
-      for (Index i = 0; i < segment->data.size(); ++i) {
-        Flow ret = runner.visit(segment->data[i]);
-        externalInterface->tableStore(
-          segment->table, offset + i, ret.getSingleValue());
+
+      Address offset =
+        (uint32_t)runner.visit(segment->offset).getSingleValue().geti32();
+
+      Table* table = wasm.getTable(segment->table);
+      if (table->imported()) {
+        auto inst = linkedInstances.at(table->module);
+        Export* tableExport = inst->wasm.getExport(table->base);
+        for (Index i = 0; i < segment->data.size(); ++i) {
+          Flow ret = runner.visit(segment->data[i]);
+          inst->externalInterface->tableStore(
+            tableExport->value, offset + i, ret.getSingleValue());
+        }
+      } else {
+        for (Index i = 0; i < segment->data.size(); ++i) {
+          Flow ret = runner.visit(segment->data[i]);
+          externalInterface->tableStore(
+            segment->table, offset + i, ret.getSingleValue());
```

Comment:
these two loops are almost identical. could the different part be branched on inside the loop? (this is not performance-sensitive code, so that's ok if it's shorter)

Reply:
Done. Although this will be replaced by a `table.init` when #3776 lands.